oxide-vpc/tests/integration_tests.rs

```diff
@@ -2218,204 +2217,356 @@ fn gateway_router_advert_reply() {
-    match reply_icmp {
-        Icmpv6Repr::Ndisc(NdiscRepr::NeighborAdvert {
-            flags,
-            target_addr,
+    let reply_icmp = Icmpv6Repr::parse(
+        &IpAddress::Ipv6(ip6.src.into()),
+        &IpAddress::Ipv6(ip6.dst.into()),
+        &reply_pkt,
+        &csum,
+    )
+    .unwrap();
+    if let Icmpv6Repr::Ndisc(NdiscRepr::NeighborAdvert {
+        flags,
+        target_addr,
+        lladdr,
+    }) = reply_icmp
+    {
+        assert_eq!(flags, na.flags);
+        assert_eq!(target_addr, na.target_addr.into());
+        assert_eq!(
             lladdr,
-        }) => {
-            assert_eq!(
-                flags,
-                NdiscNeighborFlags::ROUTER
-                    | NdiscNeighborFlags::SOLICITED
-                    | NdiscNeighborFlags::OVERRIDE
-            );
-            assert_eq!(
-                target_addr,
-                Ipv6Addr::from_eui64(&g1_cfg.gateway_mac).into()
-            );
-            assert_eq!(
-                lladdr.expect("Expected a Link-Layer Address option"),
-                RawHardwareAddress::from_bytes(&g1_cfg.gateway_mac.bytes())
-            );
-        }
-        other => {
-            panic!(
-                "Expected an ICMPv6 Neighbor Advertisement, found {:?}",
-                other
-            );
-        }
-    };
+            na.lladdr.map(|x| RawHardwareAddress::from_bytes(&x.bytes()))
+        );
+    } else {
+        panic!(
+            "Expected an ICMPv6 Neighbor Advertisement, found {:?}",
+            reply_icmp
+        );
+    }
+}
+
+// Ensure that we either Drop a Neighbor Solicitation, or generate a Neighbor
+// Advertisement with the right data, based on our defined test cases.
+#[test]
+fn test_gateway_neighbor_advert_reply() {
+    let g1_cfg = g1_cfg();
+    let v2p = Arc::new(Virt2Phys::new());
+    let mut ameta = ActionMeta::new();
+    let mut g1 = oxide_net_setup("g1_port", &g1_cfg, v2p.clone());
+    g1.port.start();
+    set_state!(g1, PortState::Running);
+    let mut pcap = PcapBuilder::new("gateway_neighbor_advert_reply.pcap");
+
+    let data = generate_solicit_test_data(&g1_cfg);
+    for d in data.into_iter().skip(1).take(1) {
+        let mut pkt = generate_neighbor_solicitation(&d.ns);
+        pcap.add_pkt(&pkt);
+        let res = g1.port.process(Out, &mut pkt, &mut ameta);
+        // Hack to deal appease borrow checker, since `res` isn't `Copy`.
+        let res_dbg = format!("{:?}", &res);
+
+        match (res, d.na) {
+            (Ok(ProcessResult::Drop { .. }), None) => {
+                // Dropped the packet, as we expected
+                continue;
+            },
+            (Ok(Hairpin(hp)), Some(na)) => {
+                assert_port!(g1);
+                println!("\n{}\n{}", d.ns, na);
+                validate_hairpin_advert(&mut pcap, hp, na);
+            }
+            _ => panic!(
+                "Generated unexpected packet from NS: {}\nResult: {}\nExpected: {:?}",
```

Comment:
Did you mean to print the NA as debug?

Reply:
I think that was me being sloppy. I implemented `Display` for these types so I didn't have to try to interpret `[u8]`s for the addresses, and then forgot to change that one. Thanks, I'll move it over now.